### ISSUE TITLE
Loss function to account for L1 loss in a GAN network

### ISSUE THREAD

**Issue by Neltherion** [fchollet/keras#4649]
Hi

I have a GAN model made up of a Generator (produces images) and a Discriminator (discriminates between real and fake generator outputs). here's the code for joining the models:

```
def generator_to_discriminator(generator, discriminator):
    model = Sequential()
    model.add(generator)
    model.add(discriminator)

    return model

generator_to_discriminator.compile(loss='bce', optimizer='adam')
```

I want to change the loss function of `generator_to_discriminator` model to also account for the `L1` loss of `generator` (which is a part of it) but a loss function only gets `y_true` and `y_pred` which are like this:

`y_true => an input image to the Generator`
`y_pred => 0 or 1 to account for the output being real or fake`

I tried writing a loss function like below but it didn't work:

```
def generator_to_discriminator_loss(y_true, y_pred):
    return binary_crossentropy(y_true, y_pred) + mae(y_true, generator.predict(y_true))
```

it gives me this error: `object of type 'TensorVariable' has no len()`.

I saw issues like #2121 but something tells me this should be easier than what has been proposed in those issues. How should I write the loss function to account for the L1 loss of the `generator` in the `generator_to_discriminator` model?

Thanks...

**Comment by bityangke** [comment]
Hi @Neltherion 
Have you solved your problem？
If you have done this， would you please share your solution？

**Comment by Neltherion** [comment]
@bityangke 
Hi
I couldn't find a way to play with the main loss function, so I ended up changing the design of the network to have 2 outputs:

1. The usual discriminator output which is a `Dense(1)` output and has a `binary_crossentropy` loss.
2. The L1 output which is just the Input of the discriminator.

Then I compiled the model like this:

`generator_to_discriminator.compile(loss=['binary_crossentropy', 'mae'], loss_weights=[1.0, 5], optimizer=Adam(lr=0.0005, beta_1=0.5))`

Now it accounts the L1 loss with the `mae` loss function and the usual discriminator loss with the `binary_crossentropy` loss.

I just did this and it's working, but the output still isn't what I wanted it to be, since when you have more than 1 output in your model, the way Keras calculates loss becomes kind of hard to understand... for example after doing what I pointed out above, Keras gives me 3 outputs whenever I call `generator_to_discriminator.train_on_batch`. the first loss seems to be the sum of the second loss (`binary_crossentropy`) and the third loss (`mae`). BUT there are cases where I can clearly see that it's not their sum and furthermore when I use weights with those losses,  I can't understand what happens in their final calculation.

I wanted a loss function like this:

```
def generator_to_discriminator_loss(y_true, y_pred):
    return binary_crossentropy(y_true, y_pred) + mae(y_true, generator.predict(y_true))
```

but I'm still not sure if what I did was the way to go... I'd sure appreciate it if someone could approve or disprove the way I did the loss in this scenario...
